pkg/controller/worker/machines.go

```diff
@@ -48,10 +53,61 @@ func (w *workerDelegate) DeployMachineClasses(ctx context.Context) error {
 			return err
 		}
 	}
-	return w.seedChartApplier.Apply(
+
+	if err := w.seedChartApplier.Apply(
 		ctx, filepath.Join(kubevirt.InternalChartsPath, "machine-class"), w.worker.Namespace, "machine-class",
 		kubernetes.Values(map[string]interface{}{"machineClasses": w.machineClasses}),
-	)
+	); err != nil {
+		return fmt.Errorf("failed to apply machine-class chart: %v", err)
+	}
+
+	dataVolumeLabels := map[string]string{
+		clusterLabel: w.worker.Namespace,
+	}
+
+	kubeconfig, err := kubevirt.GetKubeConfig(ctx, w.Client(), w.worker.Spec.SecretRef)
+	if err != nil {
+		return fmt.Errorf("failed to get the kubeconfig of the kubevirt cluster: %v", err)
+	}
+
+	for _, machineClass := range w.machineClasses {
+		var (
+			storageClassName = pointer.StringPtr(machineClass["storageClassName"].(string))
+			pvcSize          = v1.ResourceList{v1.ResourceStorage: machineClass["pvcSize"].(resource.Quantity)}
+			sourceUrl        = machineClass["sourceURL"].(string)
+			machineClassName = machineClass["name"].(string)
+		)
+
+		dataVolumeSpec := cdi.DataVolumeSpec{
+			PVC: &v1.PersistentVolumeClaimSpec{
+				StorageClassName: storageClassName,
+				AccessModes: []v1.PersistentVolumeAccessMode{
+					"ReadWriteOnce",
+				},
+				Resources: v1.ResourceRequirements{
+					Requests: pvcSize,
+				},
+			},
+			Source: cdi.DataVolumeSource{
+				HTTP: &cdi.DataVolumeSourceHTTP{
+					URL: sourceUrl,
+				},
+			},
+		}
+
+		dataVolume := &cdi.DataVolume{}
+		dataVolume.Name = machineClassName
+		dataVolume.Spec = dataVolumeSpec
+		dataVolume.Labels = dataVolumeLabels
+
+		if err := retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+			return w.dataVolumeManager.CreateVolume(ctx, kubeconfig, dataVolume)
+		}); err != nil {
+			return fmt.Errorf("failed to create data volume for machine class %s: %v", machineClassName, err)
+		}
+	}
+
```

Comment:
I think that after creating or updating all the data volumes, you should wait for the import to all of them to finish before you exit from this method (you may run into timeouts though, this should be tested). Otherwise, when the VMs are created, they would not be able to use the pre-allocated data volumes. I could also be wrong here, have you tested this?

Reply:
Not really, Kubevirt operator would actually mark that, there is no datavolume is there, thus it will keep trying to clone the referenced the datavolume until it's there. In other words, the creation of the vm is not dependent on the datavolume, the vmi is, thus creation of the vm won't fail and once the dv is there, the operator will simply pick up the dv, cloning it and attaching it to the vmi and run it.

Reply:
Thanks for explaining, sounds good.